src/interpolation/details/ArborX_InterpDetailsPolynomialBasis.hpp

```diff
@@ -0,0 +1,152 @@
+{
+
+// Polynomial basis is computed in the same way as the following example
+// diagrams. Each cell is the product of the variable times everything on the
+// left cell and above.
+//
+// For example, the "degree 3 / variable y" has size 3 because it multiplies
+// every element of "degree 2 / variable y" and "degree 2 / variable x".
+//
+//   Quadratic |  3D  ||        Cubic      |  2D
+// ------------+----- || ------------------+-----
+//    degree   | var  ||       degree      | var
+// ---+---+----+      || ---+---+----+-----+
+//  0 | 1 |  2 |      ||  0 | 1 |  2 |  3  |
+// ---+---+----+      || ---+---+----+-----+
+// ---+---+----+      || ---+---+----+-----+
+//  1 |   |    |      ||  1 |   |    |     |
+// ---+---+----+----- || ---+---+----+-----+-----
+//    | x |    |  x   ||    | x |    |     |  x
+//    |   | xx |      ||    |   | xx |     |
+//    +---+----+----- ||    |   |    | xxx |
+//    | y |    |  y   ||    +---+----+-----+-----
+//    |   | xy |      ||    | y |    |     |  y
+//    |   | yy |      ||    |   | xy |     |
+//    +---+----+----- ||    |   | yy |     |
+//    | z |    |  z   ||    |   |    | xxy |
+//    |   | xz |      ||    |   |    | xyy |
+//    |   | yz |      ||    |   |    | yyy |
+//    |   | zz |      ||
+
+// The size of each cell is 1 if it is at degree 1 or at variable x
+// And the sum of the cells' size to the left and above. (1 alone is not
+// counted). This essentially creates Pascal's triangle where line n corresponds
+// to the "dim + deg = n"-th diagonal
+template <std::size_t Dim, std::size_t Deg>
+KOKKOS_FUNCTION constexpr auto polynomialBasisCellSizes()
+{
+  static_assert(Deg != 0 && Dim != 0,
+                "Unable to compute cell sizes for a constant polynomial basis");
+
+  struct
+  {
+    std::size_t arr[Deg][Dim]{};
+  } result;
+  auto &arr = result.arr;
+
+  for (std::size_t dim = 0; dim < Dim; dim++)
+    arr[0][dim] = 1;
+
+  for (std::size_t deg = 0; deg < Deg; deg++)
+    arr[deg][0] = 1;
+
+  for (std::size_t deg = 1; deg < Deg; deg++)
+    for (std::size_t dim = 1; dim < Dim; dim++)
+      arr[deg][dim] = arr[deg - 1][dim] + arr[deg][dim - 1];
+
+  return result;
+}
+
+// This returns the size of the polynomial basis, which is the sum of all the
+// cells' sizes and 1.
+template <std::size_t Dim, std::size_t Deg>
+KOKKOS_FUNCTION constexpr std::size_t polynomialBasisSize()
+{
+  if constexpr (Deg != 0 && Dim != 0)
+  {
+    auto [arr] = polynomialBasisCellSizes<Dim, Deg>();
+    std::size_t size = 1;
+
+    for (std::size_t deg = 0; deg < Deg; deg++)
+      for (std::size_t dim = 0; dim < Dim; dim++)
+        size += arr[deg][dim];
+
+    return size;
+  }
+  else
+  {
+    return 1;
+  }
+}
+
+// This builds the array as described above
+template <std::size_t Dim, std::size_t Deg, typename Point>
+KOKKOS_FUNCTION auto polynomialBasis(Point const &p)
+{
+  using value_t = std::decay_t<decltype(p[0])>;
+  Kokkos::Array<value_t, polynomialBasisSize<Dim, Deg>()> arr{};
```

Comment:
Why don't we return a 2D C-style array? How do you map back from the linear index to `DIm` and `Degree`?

Reply:
> Why don't we return a 2D C-style array?

First, you cannot return a C-style array. Second, it would not really work to have a 2D array as the rows and the columns have different sizes.

> How do you map back from the linear index to `Dim` and `Degree`?

This function simply flattens the 2D grid of cells in a row-major way. So, if you know `Dim` and `Deg`, you can get the cell sizes and reverse the process.

Reply:
Fair enough.